flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/util/TableResultUtils.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.DataStreamSink;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.table.runtime.types.LogicalTypeDataTypeConverter;
+import org.apache.flink.table.runtime.types.TypeInfoDataTypeConverter;
+import org.apache.flink.table.sinks.AppendStreamTableSink;
+import org.apache.flink.table.sinks.TableSink;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.TimestampKind;
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.AbstractID;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A collection of utilities for fetching table results.
+ *
+ * <p>NOTE: Methods in this utility class are experimental and can only be used for demonstration or testing
+ * small table results. Please DO NOT use them in production or on large tables.
+ *
+ * <p>We ought to put this class in the api module, but as converting high precision timestamp data type
+ * to type information is not possible in flink planner, we have to put this class in blink planner.
+ */
+@Experimental
+public class TableResultUtils {
+
+	/**
+	 * Convert Flink table to Java list.
+	 * This method is only applicable for small batch jobs and small finite append only stream jobs.
+	 *
+	 * @param table		Flink table to convert
+	 * @return			Converted Java list
+	 */
+	@SuppressWarnings("unchecked")
+	public static List<Row> tableResultToList(Table table) throws Exception {
+		final TableEnvironment tEnv = ((TableImpl) table).getTableEnvironment();
+		final String id = new AbstractID().toString();
+		final TableSchema schema = buildNewTableSchema(table);
+		final DataType rowDataType = schema.toRowDataType();
+		final TypeSerializer<Row> serializer =
+			(TypeSerializer<Row>) TypeInfoDataTypeConverter
+				.fromDataTypeToTypeInfo(rowDataType)
+				.createSerializer(new ExecutionConfig());
+		final Utils.CollectHelper<Row> outputFormat = new Utils.CollectHelper<>(id, serializer);
+		final TableResultSink sink = new TableResultSink(schema, outputFormat);
+
+		final String sinkName = "tableResultSink_" + id;
+		final String jobName = "tableResultToList_" + id;
+
+		tEnv.registerTableSink(sinkName, sink);
+		tEnv.insertInto(sinkName, table);
+
+		final JobExecutionResult executionResult = tEnv.execute(jobName);
+		final ArrayList<byte[]> accResult = executionResult.getAccumulatorResult(id);
+		final List<Row> deserializedList = SerializedListAccumulator.deserializeList(accResult, serializer);
+
+		tEnv.dropTemporaryTable(sinkName);
+		return deserializedList;
+	}
+
+	private static TableSchema buildNewTableSchema(Table table) {
```

Comment:
Could you add some comment to explain why we need to build new table schema here ?